standalone/src/test/java/com/tngtech/keycloakmock/standalone/ServerTest.java

```diff
@@ -121,12 +135,113 @@ void mock_server_properly_returns_resources(
     RestAssured.given()
         .when()
         .get("http://localhost:8001/i-do-not-exist")
         .then()
         .assertThat()
         .statusCode(404);
   }
+
+  @Test
+  void mock_server_login_works() throws Exception {
+    JwtParser jwtParser = Jwts.parserBuilder().setSigningKey(loadKey()).build();
+    server = new Server(8001, false, Collections.emptyList());
+
+    // open login page to create session (implicit flow)
+    String callbackUrl =
+        RestAssured.given()
+            .when()
+            .get(
+                "http://localhost:8001/auth/realms/realm/protocol/openid-connect/auth?client_id=client&state=state&nonce=nonce&redirect_uri=redirect_uri&response_type=id_token")
+            .then()
+            .assertThat()
+            .statusCode(200)
+            .extract()
+            .htmlPath()
+            .getNode("html")
+            .getNode("body")
+            .getNode("form")
+            .getAttribute("action");
+
+    // simulate login POST and get id_token in location response header
+    ExtractableResponse<Response> extractableResponse =
+        RestAssured.given()
+            .config(config().redirect(redirectConfig().followRedirects(false)))
+            .when()
+            .formParam("username", "username")
+            .formParam("password", "role1,role2,role3")
+            .post(callbackUrl)
+            .then()
+            .assertThat()
+            .statusCode(302)
+            .extract();
+    String location = extractableResponse.header("location");
+    Cookie keycloakSession = extractableResponse.detailedCookie("KEYCLOAK_SESSION");
+
+    assertThat(location).contains("redirect_uri#state=state", "&session_state=", "id_token=");
+
+    String token = location.split("id_token=")[1];
+
+    Jws<Claims> jwt = jwtParser.parseClaimsJws(token);
+
+    assertThat(jwt.getBody().getIssuer()).isEqualTo("http://localhost:8001/auth/realms/realm");
+
+    TokenConfig tokenConfig = TokenConfig.aTokenConfig().withSourceToken(token).build();
+
+    assertThat(tokenConfig.getPreferredUsername()).isEqualTo("username");
+    assertThat(tokenConfig.getRealmAccess().getRoles())
+        .containsExactlyInAnyOrder("role1", "role2", "role3");
+    assertThat(tokenConfig.getClaims()).containsEntry("nonce", "nonce");
+
+    // subsequent request to login page with session cookie set will immediately return
+    String locationFromSession =
+        RestAssured.given()
+            .config(config().redirect(redirectConfig().followRedirects(false)))
+            .when()
+            .cookie(keycloakSession)
+            .get(
+                "http://localhost:8001/auth/realms/realm/protocol/openid-connect/auth?client_id=client2&state=state2&nonce=nonce2&redirect_uri=redirect_uri2&response_type=id_token")
+            .then()
+            .assertThat()
+            .statusCode(302)
+            .extract()
+            .header("location");
+
+    assertThat(locationFromSession)
```

Comment:
Looks like a view lines could be extracted here to avoid duplication ...

Reply:
I'll have a look.